### PR TITLE
Add new InstallBundle Method to D-Bus API

### PR DESCRIPTION
**Pull request by ejoerns**
In the past, there were a lot of cases where we ran into the limitation of the `Install` method of the D-Bus `Installer` interface. These were all caused by the missing ability to pass additional arguments to the install call.

As we cannot break existing setups, we cannot simply change the method signature. Instead, we need to introduce an alternative method.

Here it is: ``InstallBundle sa{sv}``

Where the new generic dict of string-variant (``a{sv}``) allows passing and handling all kind of current and future parameters without changing the interface again.

This initial series includes support for one argument that was already implemented for the non-D-Bus case: ``--ignore-compatible``. It allows forcing the installation of a bundle on a per-installation-call base

Reference: #248

### CODE REVIEW

**Comment by jluebbe** [comment]
I've pushed a rebase onto master. The the conflicts were caused by the network mode removal.

**Comment by ejoerns** [comment]
We also have tests for this and it remained in RFC phase for some days. Thus we can remove the RFC label now and make it a normal PR ready for being reviewed and merged.

File: include/context.h [hunk 1]
```diff
@@ -124,4 +122,4 @@ void r_context_register_progress_callback(progress_callback progress_cb);
 
 RaucContext *r_context_conf(void);
 const RaucContext *r_context(void);
-void r_context_clean(void);
\ No newline at end of file
+void r_context_clean(void);
```
Comment (michaelolbrich) [inline include/context.h]: unrelated change.

File: src/service.c [hunk 2]
```diff
@@ -70,6 +71,11 @@ static gboolean r_on_handle_install_bundle(
 	args->notify = service_install_notify;
 	args->cleanup = service_install_cleanup;
 
+	g_variant_dict_init(&dict, arg_args);
+	g_variant_dict_lookup(&dict, "ignore-compatible", "b", &args->ignore_compatible);
+	arg_args = g_variant_dict_end(&dict);
+	g_variant_unref(arg_args);
```
Comment (michaelolbrich) [inline src/service.c]: This is a bit strange. I'm not sure if `g_variant_dict_init()` consumes the ref for `arg_args` or not, but either way, `g_variant_dict_clear()` should be used here and maybe `g_variant_unref()` if necessary.

File: test/service.c [hunk 3]
```diff
@@ -203,8 +203,17 @@ static void service_test_install(ServiceFixture *fixture, gconstpointer user_dat
...
 
 	/* Actually install bundle */
-	ret = r_installer_call_install_sync(installer, bundlepath, NULL,
-			&error);
+	if (deprecated) {
+		ret = r_installer_call_install_sync(installer, bundlepath, NULL,
+				&error);
+	} else {
+		GVariantDict dict;
+		GVariant *install_args = NULL;
```
Comment (michaelolbrich) [inline test/service.c]: no need to initialize `install_args`.

File: src/service.c (hunk 2 above)
Comment (michaelolbrich) [reply src/service.c]: Also, you might want to change the commit order: Before this change, `arg_args` was leaked.

File: include/context.h (hunk 1 above)
Comment (ejoerns) [reply include/context.h]: Appears to be a line at end-of-file change done by my editor, will remove it.

File: test/service.c (hunk 3 above)
Comment (ejoerns) [reply test/service.c]: I'll just drop it

File: src/service.c (hunk 2 above)
Comment (ejoerns) [reply src/service.c]: Right, `g_variant_dict_clear()` should do it here then there's no need to unref (and abuse `arg_args`).

About the `arg_args` leakage: This is passed from outside and thus I would expect the caller to free it.

File: src/main.c [hunk 4]
```diff
@@ -242,12 +248,15 @@ static gboolean install_start(int argc, char **argv)
...
-		if (!r_installer_call_install_sync(installer, args->name, NULL,
+		if (!r_installer_call_install_bundle_sync(installer, args->name, arg_args, NULL,
 				&error)) {
 			g_printerr("Failed %s\n", error->message);
 			g_error_free(error);
+			g_variant_unref(arg_args);
 			goto out_loop;
 		}
+
+		g_variant_unref(arg_args);
```
Comment (michaelolbrich) [inline src/main.c]: I'm pretty sure `arg_args` leaks in the error case. Maybe use a cleanup handler?

File: src/service.c (hunk 2 above)
Comment (michaelolbrich) [reply src/service.c]: That depends on how the API is defined. And the caller is generated gdbus code, right?

File: src/main.c (hunk 4 above)
Comment (ejoerns) [reply src/main.c]: Ouh, I've missed the first error handling block, only added it to the second one.

Comment (jluebbe) [reply src/main.c]: I've refactored this.

File: src/install.c [hunk 5]
```diff
@@ -1144,7 +1144,6 @@ void install_args_free(RaucInstallArgs *args)
 {
 	g_free(args->name);
 	g_mutex_clear(&args->status_mutex);
-	g_assert_cmpint(args->status_result, >=, 0);
```
Comment (jluebbe) [inline src/install.c:1147]: Why did this fail only now? In src/main.c, we've been freeing install args for a long time.

Comment (ejoerns) [reply src/install.c:1147]: Ok, I had a closer look at this now as I knew I came across this but did not recap where actually.

The potential problem I referred to is that `install_args_new()` initializes `status_result` to `-2`.
Thus if `install_args_free()` gets called before some instance changed this value, we will end up in the assertion error. But maybe, yes, these probably are the exact cases that we wanted to find with the assertion...

The only two points I dug out where this could have been an issue are:

https://github.com/rauc/rauc/blob/master/src/service.c#L63

and

https://github.com/rauc/rauc/blob/master/src/install.c#L1159

As they potentially return without having the args touched.

BUT, as we do

https://github.com/rauc/rauc/blob/master/src/service.c#L80

(using only `g_free()` on the complex initialized type).This will never trigger.

IMHO this should call `install_args_free()` here.
Also setting `args = NULL` above seems to be a bit unsafe but should not make any issues if we properly cleared args before.

Comment (jluebbe) [reply src/install.c]: This should be solved now.
